packages/apollo-engine-reporting/src/extension.ts

```diff
@@ -455,3 +428,153 @@ function defaultGenerateClientInfo({ request }: GraphQLRequestContext) {
     return {};
   }
 }
+
+// Creates trace details from request variables, given a specification for modifying
+// values of private or sensitive variables.
+// The details include the variables in the request, TODO(helen): and the modifier type.
+// If sendVariableValues is {safelistAll: bool} or {exceptVariableNames: Array}, it will act similarly to
+// to the to-be-deprecated options.privateVariables, except that the redacted variable
+// names will still be visible in the UI when 'true.'
+// If sendVariableValues is null, we default to the safeListAll = false case.
+export function makeTraceDetails(
+  variables: Record<string, any>,
+  sendVariableValues?: VariableValueOptions,
+  operationString?: string,
+): Trace.Details {
+  const details = new Trace.Details();
+  const variablesToRecord = (() => {
+    if (sendVariableValues && 'valueModifier' in sendVariableValues) {
+      // Custom function to allow user to specify what variablesJson will look like
+      const originalKeys = Object.keys(variables);
+      const modifiedVariables = sendVariableValues.valueModifier({
+        variables: variables,
+        operationString: operationString,
+      });
+      return cleanModifiedVariables(originalKeys, modifiedVariables);
+    } else {
+      return variables;
+    }
+  })();
+
+  // Note: we explicitly do *not* include the details.rawQuery field. The
+  // Engine web app currently does nothing with this other than store it in
+  // the database and offer it up via its GraphQL API, and sending it means
+  // that using calculateSignature to hide sensitive data in the query
+  // string is ineffective.
+  Object.keys(variablesToRecord).forEach(name => {
+    if (
+      sendVariableValues == null ||
+      ('safelistAll' in sendVariableValues &&
+        !sendVariableValues.safelistAll) ||
+      ('exceptVariableNames' in sendVariableValues &&
+        // We assume that most users will have only a few private variables,
+        // or will just set privateVariables to true; we can change this
+        // linear-time operation if it causes real performance issues.
+        sendVariableValues.exceptVariableNames.includes(name))
+    ) {
+      // Special case for private variables. Note that this is a different
+      // representation from a variable containing the empty string, as that
+      // will be sent as '""'.
+      details.variablesJson![name] = '';
+    } else {
+      try {
+        details.variablesJson![name] =
+          variablesToRecord[name] == null
```

Comment:
`null` is actually a totally legitimate GraphQL variable value which should probably be sent over as `null`. I'd differentiate that from `undefined` (which generally in JS as an object value means "don't send me when serializing", ie `JSON.stringify({a: undefined}) === "{}`).